File: src/algorithms/cpp/quick-sort.cpp
#include <iostream>
#include <vector>

using namespace std;

void emitCompare(int i, int j) {
    cout << "{\"type\":\"ARRAY_COMPARE\",\"indices\":[" << i << "," << j << "]}" << endl;
}
void emitSwap(int i, int j, int vi, int vj) {
    cout << "{\"type\":\"ARRAY_SWAP\",\"indices\":[" << i << "," << j << "],\"values\":[" << vi << "," << vj << "]}" << endl;
}
void emitInit(const vector<int>& arr) {
    cout << "{\"type\":\"INIT\",\"array\":[";
    for (int i = 0; i < (int)arr.size(); i++) {
        if (i > 0) cout << ",";
        cout << arr[i];
    }
    cout << "]}" << endl;
}

int partition(vector<int>& arr, int low, int high) {
  int pivot = arr[high];
  int i = low - 1;
  for (int j = low; j < high; j++) {
    emitCompare(j, high);
    if (arr[j] < pivot) {
      i++;
      if (i != j) {
        emitSwap(i, j, arr[j], arr[i]);
        swap(arr[i], arr[j]);
      }
    }
  }
  if (i + 1 != high) {
    emitSwap(i + 1, high, arr[high], arr[i + 1]);
    swap(arr[i + 1], arr[high]);
  }
  return i + 1;
}

void quickSort(vector<int>& arr, int low, int high) {
  if (low < high) {
    int pi = partition(arr, low, high);
    quickSort(arr, low, pi - 1);
    quickSort(arr, pi + 1, high);
  }
}

int main() {
    vector<int> arr = {29, 10, 14, 37, 14, 25, 1, 31, 9, 22};
    emitInit(arr);
    quickSort(arr, 0, arr.size() - 1);
    return 0;
}

File: src/algorithms/python/quick-sort.py
import json

def emit(event_type, **kwargs):
    event = {"type": event_type}
    event.update(kwargs)
    print(json.dumps(event))

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        emit("ARRAY_COMPARE", indices=[j, high])
        if arr[j] < pivot:
            i += 1
            if i != j:
                emit("ARRAY_SWAP", indices=[i, j], values=[arr[j], arr[i]])
                arr[i], arr[j] = arr[j], arr[i]
    if i + 1 != high:
        emit("ARRAY_SWAP", indices=[i + 1, high], values=[arr[high], arr[i + 1]])
        arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

if __name__ == "__main__":
    arr = [29, 10, 14, 37, 14, 25, 1, 31, 9, 22]
    emit("INIT", array=arr[:])
    quick_sort(arr, 0, len(arr) - 1)
